File: FindTheSmallest/FindTheSmallest_test.go
package findthesmallest

import (
	"reflect"
	"testing"
)

func TestSmallest(t *testing.T) {
	testCases := []struct {
		input    int64
		expected []int64
	}{
		{261235, []int64{126235, 2, 0}},
		{209917, []int64{29917, 0, 1}},
		{199819884756, []int64{119989884756, 4, 0}},
		{5190930055291, []int64{519093055291, 6, 0}},
		{60496893958362, []int64{4696893958362, 0, 2}},
	}

	for _, c := range testCases {
		got := Smallest(c.input)
		if !reflect.DeepEqual(got, c.expected) {
			t.Errorf("TestSmallest: input: %v, expected: %v, got: %v", c.input, c.expected, got)
		}
	}
}

func TestExchange(t *testing.T) {
	testCases := []struct {
		arr      []int
		from     int
		to       int
		expected []int
	}{
		{[]int{0, 1, 2, 3, 4, 5}, 3, 1, []int{0, 3, 1, 2, 4, 5}},
		{[]int{0, 1, 2, 3, 4, 5}, 1, 3, []int{0, 2, 3, 1, 4, 5}},
	}

	for _, testCase := range testCases {
		got := move(testCase.arr, testCase.from, testCase.to)
		if !reflect.DeepEqual(got, testCase.expected) {
			t.Errorf("TestExchange: got %v (%d => %d), expected: %v, got: %v",
				testCase.arr, testCase.from, testCase.to,
				testCase.expected, got)
		}
	}
}

func TestSliceToInt64(t *testing.T) {
	testCases := []struct {
		input    []int
		expected int64
	}{
		{[]int{1, 2, 1, 1, 1}, 12111},
		{[]int{1, 2, 3, 4, 5, 6}, 123456},
		{[]int{6, 5, 4, 3, 2, 1}, 654321},
		{[]int{6, 5, 4, 7, 5, 4, 8}, 6547548},
	}

	for _, testCase := range testCases {
		got := sliceToInt64(testCase.input)
		if got != testCase.expected {
			t.Errorf("TestSliceToInt64: expected: %v, got: %v", testCase.expected, got)
		}
	}
}

func BenchmarkSmallest(b *testing.B) {
	for i := 0; i < b.N; i++ {
		Smallest(5275027536092356050)
	}
}

func BenchmarkSmallest2(b *testing.B) {
	for i := 0; i < b.N; i++ {
		Smallest2(5275027536092356050)
	}
}

File: FindTheSmallest/FindTheSmallest_test.js
var assert = require('assert')
var findTheSmallest = require('./FindTheSmallest')

describe('sallest', () => {
    let testCases = [
        { input: 261235, expected: { value: 126235, from: 2, to: 0 } },
        { input: 209917, expected: { value: 29917, from: 0, to: 1 } },
        { input: 199819884756, expected: { value: 119989884756, from: 4, to: 0 } },
        { input: 5190930055291, expected: { value: 519093055291, from: 6, to: 0 } },
        { input: 60496893958362, expected: { value: 4696893958362, from: 0, to: 2 } },
        { input: 111411111, expected: { value: 111111114, from: 3, to: 8 } },
    ]

    testCases.forEach((testCase) => {
        let got = findTheSmallest.smallest(testCase.input);
        it(`Input: ${testCase.input}`, () => assert.deepEqual(got, testCase.expected))
    })
})

describe('numToArray', () => {
    let testCases = [
        { input: 123456, expected: [1, 2, 3, 4, 5, 6] },
        { input: 0, expected: [] },
        { input: -1, expected: [] },
    ]

    testCases.forEach((testCase) => {
        let got = findTheSmallest.numToArray(testCase.input);

        it(`Input: ${testCase.input}`, () => {
            assert.deepEqual(got, testCase.expected)
        })
    })
})

describe('arrayToNum', () => {
    let testCases = [
        { expected: 123456, input: [1, 2, 3, 4, 5, 6] },
        { expected: 0, input: [] },
    ]

    testCases.forEach((testCase) => {
        let got = findTheSmallest.arrayToNum(testCase.input);

        it(`Input: ${testCase.input}`, () => {
            assert.equal(got, testCase.expected)
        })
    })
})


describe('move', () => {
    let testCases = [
        { input: [1, 2, 3, 4, 5, 6], from: 2, to: 0, expected: [3, 1, 2, 4, 5, 6], },
        { input: [1, 2, 3, 4, 5, 6], from: 2, to: 4, expected: [1, 2, 4, 5, 3, 6], },
    ]

    testCases.forEach((testCase) => {
        let got = findTheSmallest.move(testCase.input, testCase.from, testCase.to);

        it(`Input: ${testCase.input}`, () => {
            assert.deepEqual(got, testCase.expected)
        })
    })
})